File: design_patterns_study/230407_factory/go/abstractFactory/factory/concrete.go
package factory

import "abstractfactory/polygon"

// build 3 structs each with a method that returns the Polygon Interface
// also build 1 interface that capture all those three structs

type FactoryColorPolygon interface {
	Create(numSides int) polygon.Polygon
}

type FactoryColorPolygonBlue struct{}

func (f FactoryColorPolygonBlue) Create(numSides int) polygon.Polygon {
	if numSides == 3 {
		return polygon.BlueTriangle{}
	} else if numSides == 4 {
		return polygon.BlueRectangle{}
	} else if numSides == 5 {
		return polygon.BluePentagon{}
	} else {
		return nil
	}
}

type FactoryColorPolygonRed struct{}

func (f FactoryColorPolygonRed) Create(numSides int) polygon.Polygon {
	if numSides == 3 {
		return polygon.RedTriangle{}
	} else if numSides == 4 {
		return polygon.RedRectangle{}
	} else if numSides == 5 {
		return polygon.RedPentagon{}
	} else {
		return nil
	}
}

type FactoryColorPolygonGreen struct{}

func (f FactoryColorPolygonGreen) Create(numSides int) polygon.Polygon {
	if numSides == 3 {
		return polygon.GreenTriangle{}
	} else if numSides == 4 {
		return polygon.GreenRectangle{}
	} else if numSides == 5 {
		return polygon.GreenPentagon{}
	} else {
		return nil
	}
}

File: design_patterns_study/230407_factory/python/abstract_factory/concrete.py
from dataclasses import dataclass
from abc import ABC, abstractmethod
import polygon


class FactoryPolygonColor(ABC):
    @abstractmethod
    def create(self, num_sides: int) -> polygon.Polygon:
        pass

@dataclass
class FactoryPolygonBlue(FactoryPolygonColor):
    def create(self, num_sides: int) -> polygon.Polygon:
        if num_sides == 3:
            return polygon.BlueTriangle()
        elif num_sides == 4:
            return polygon.BlueRectangle()
        elif num_sides == 5:
            return polygon.BluePentagon()
        else:
            raise Exception("Sorry, we've got a problem...")


@dataclass
class FactoryPolygonRed(FactoryPolygonColor):
    def create(self, num_sides: int) -> polygon.Polygon:
        if num_sides == 3:
            return polygon.RedTriangle()
        elif num_sides == 4:
            return polygon.RedRectangle()
        elif num_sides == 5:
            return polygon.RedPentagon()
        else:
            raise Exception("Sorry, we've got a problem...")


@dataclass
class FactoryPolygonGreen(FactoryPolygonColor):
    def create(self, num_sides: int) -> polygon.Polygon:
        if num_sides == 3:
            return polygon.GreenTriangle()
        elif num_sides == 4:
            return polygon.GreenRectangle()
        elif num_sides == 5:
            return polygon.GreenPentagon()
        else:
            raise Exception("Sorry, we've got a problem...")
